src/expr/impl/src/udf/quickjs.rs

```diff
@@ -85,37 +107,39 @@ struct QuickJsFunction {
 #[async_trait::async_trait]
 impl UdfImpl for QuickJsFunction {
     async fn call(&self, input: &RecordBatch) -> Result<RecordBatch> {
-        self.runtime.call(&self.identifier, input)
+        self.runtime.call(&self.identifier, input).await
     }
 
     async fn call_table_function<'a>(
         &'a self,
         input: &'a RecordBatch,
     ) -> Result<BoxStream<'a, Result<RecordBatch>>> {
-        self.runtime
-            .call_table_function(&self.identifier, input, 1024)
-            .map(|s| futures_util::stream::iter(s).boxed())
+        let iter = self
+            .runtime
+            .call_table_function(&self.identifier, input, 1024)?;
+        Ok(Box::pin(iter))
     }
 
-    fn call_agg_create_state(&self) -> Result<ArrayRef> {
-        self.runtime.create_state(&self.identifier)
+    async fn call_agg_create_state(&self) -> Result<ArrayRef> {
+        self.runtime.create_state(&self.identifier).await
     }
 
-    fn call_agg_accumulate_or_retract(
+    async fn call_agg_accumulate_or_retract(
         &self,
         state: &ArrayRef,
         ops: &BooleanArray,
         input: &RecordBatch,
     ) -> Result<ArrayRef> {
         self.runtime
             .accumulate_or_retract(&self.identifier, state, ops, input)
+            .await
     }
 
-    fn call_agg_finish(&self, state: &ArrayRef) -> Result<ArrayRef> {
-        self.runtime.finish(&self.identifier, state)
+    async fn call_agg_finish(&self, state: &ArrayRef) -> Result<ArrayRef> {
+        self.runtime.finish(&self.identifier, state).await
     }
 
-    fn memory_usage(&self) -> usize {
-        self.runtime.memory_usage().malloc_size as usize
+    async fn memory_usage(&self) -> usize {
+        self.runtime.inner().memory_usage().await.malloc_size as usize
     }
```

Comment:
I'm wondering if we can do `self.runtime.inner().memory_usage().now_or_never().unwrap_or_default()` here. Making `fn memory_usage` async looks strange to me.

Reply:
Yeah, it's stupid. It's just for `self.inner.lock().await`, which is introduced by the Rust binding (`rquickjs`) rather than `quickjs`.

For the same reason of https://github.com/risingwavelabs/risingwave/pull/20403#discussion_r1948715142, I tend to use `block_on` here; and because JS Runtime is never shared, the function must never block.